### ISSUE TITLE
Feature request: faster dictionary-based decoding

### ISSUE THREAD

**Issue by neeraj-kumar** [ilanschnell/bitarray#102]
Currently, it looks everytime `decode()` or `iterdecode()` is called, the value -> bits dict is converted to a binary tree. Is there a way to get a handle to that tree to avoid repeatedly constructing it on every decode?

**Comment by ilanschnell** [comment]
Yes, the binary tree is constructed from the code dictionary every time `decode()` or `iterdecode()` is called.  There is currently no way to cache this (C) object.  However, the construction of this tree is very fasted, at least its construction time will be neglectable compared to the decode time for large arrays.  I'm curious what your use-case is, and why you have to call the decode function many times.  Maybe you can provide some sample code.  Thanks!

**Comment by neeraj-kumar** [comment]
Hi, thanks for the prompt reply!

I have a large database of values I'm storing in and LMDB database, where
each value is a list of bit-encoded indices (using a custom dictionary
code). Each individual list is fairly small (2-10 items), but I have a huge
number of keys (100s of millions). As I update this database, I decode the
bit-encoded indices using the bitarrray `decode()` function, add new values
to the list, and re-encode it. Because each list is pretty small but there
are tons of them, the decode is a non-trivial portion of the compute.

**Comment by ilanschnell** [comment]
I see basically two options:

1.) Do the decoding in Python:
Doing the decoding in Python is much slower, but it would allow storing the binary tree easily.  There is an example which includes a speed comparison https://github.com/ilanschnell/bitarray/blob/master/examples/huffman/decoding.py (C vs. Python).

2.) Store the binary tree (C object):
This would require changing the C implementation.  For example, one could add a function to bitarray called `set_decoding(codedict)` which sets a the binary tree.  Then, calling `.decode(None)` or `.iterdecode(None)` (with `None` instead of the code dict) would use the global binary tree.

**Comment by ilanschnell** [comment]
I've been thinking about this problem some more.  Is your code really a variable length prefix code?  Or do all bitarrays in the code have the same length?

If all bitarrays have the same length, then it would actually be more efficient to create a reverse code dictionary (mapping frozenbitarrays to symbols) and do the decoding by a hash lookup.  The following program shows an example of what I mean:
```
import random
from time import time

from bitarray import bitarray, frozenbitarray
from bitarray.util import int2ba

N = 1000   # number of integers in test array
M = 20  # bits per key
codedict = {i: int2ba(i, M) for i in range(1 << M)}

# list with N random integers we want to encode
r = [random.randint(0, (1 << M) - 1) for i in range(N)]

a = bitarray()
a.encode(codedict, r)

t0 = time()
s = a.decode(codedict)
print(time() - t0)
assert s == r

# create reverse code dictionary
rev = {frozenbitarray(v): k for k, v in codedict.items()}

t0 = time()
s = [rev[frozenbitarray(a[i:i + M])] for i in range(0, N * M, M)]
print(time() - t0)
assert s == r
```
The output of running this program shows that using the reverse dictionary to do the fix length decoding is almost 100 times faster.

**Comment by neeraj-kumar** [comment]
I'm using elias-delta coding, so it is a variable length code -- but
identical for all items.

**Comment by ilanschnell** [comment]
@neeraj-kumar I just merged #103 which adds the `decodetree` object.
This new feature will be part of the bitarray 1.6.0 release (which I'm planning to release within a few weeks).

**Comment by neeraj-kumar** [comment]
Awesome, thanks!

**Comment by ilanschnell** [comment]
I just released bitarray 1.6.0, which includes the decodetree object.  Have fun with it!
